src/code/sys_math3d.c

```diff
@@ -0,0 +1,1692 @@
+ * Calculates the normal vector to a surface with sides `vb` - `va` and `vc` - `va`
+ * outputs the normal to `normal`
+ */
+void Math3D_SurfaceNorm(Vec3f* va, Vec3f* vb, Vec3f* vc, Vec3f* normal) {
+    static Vec3f abDiff;
+    static Vec3f acDiff;
+
+    Math_Vec3f_Diff(vb, va, &abDiff);
+    Math_Vec3f_Diff(vc, va, &acDiff);
+    Math3D_Vec3f_Cross(&abDiff, &acDiff, normal);
+}
+
+s32 func_800CB88C(Vec3f* arg0, Vec3f* arg1, Vec3f* arg2) {
+    s32 ret = 0;
+
+    if (arg2->x < arg0->x) {
+        ret = 1;
+    }
+
+    if (arg0->x < arg1->x) {
+        ret |= 2;
+    }
+
+    if (arg2->y < arg0->y) {
+        ret |= 4;
+    }
+
+    if (arg0->y < arg1->y) {
+        ret |= 8;
+    }
+
+    if (arg2->z < arg0->z) {
+        ret |= 0x10;
+    }
+
+    if (arg0->z < arg1->z) {
+        ret |= 0x20;
+    }
+    return ret;
+}
+
+#ifdef NON_MATCHING
+s32 func_800CB934(Vec3f* arg0, Vec3f* arg1, Vec3f* arg2) {
+    s32 ret = 0;
+
+    if ((arg2->y - arg1->x) < (arg0->y - arg0->x)) {
+        ret = 1;
+    }
+
+    if ((arg0->y - arg0->x) < (arg1->y - arg2->x)) {
+        ret |= 2;
+    }
+
+    if ((arg2->x + arg2->y) < (arg0->x + arg0->y)) {
+        ret |= 4;
+    }
+
+    if ((arg0->x + arg0->y) < (arg1->x + arg1->y)) {
+        ret |= 8;
+    }
+
+    if ((-arg1->z + arg2->y) < (-arg0->z + arg0->y)) {
+        ret |= 0x10;
+    }
+
+    if ((-arg0->z + arg0->y) < (arg1->y - arg2->z)) {
+        ret |= 0x20;
+    }
+
+    if ((arg2->z + arg2->y) < (arg0->z + arg0->y)) {
+        ret |= 0x40;
+    }
+
+    if ((arg0->z + arg0->y) < (arg1->z + arg1->y)) {
+        ret |= 0x80;
+    }
+
+    if ((-arg1->z + arg2->x) < (-arg0->z + arg0->x)) {
+        ret |= 0x100;
+    }
+
+    if ((-arg0->z + arg0->x) < (-arg2->z + arg1->x)) {
+        ret |= 0x200;
+    }
+
+    if ((arg2->z + arg2->x) < (arg0->z + arg0->x)) {
+        ret |= 0x400;
+    }
+
+    if ((arg0->z + arg0->x) < (arg1->z + arg1->x)) {
+        ret |= 0x800;
+    }
+    return ret;
+}
+#else
+#pragma GLOBAL_ASM("asm/non_matchings/code/sys_math3d/func_800CB934.s")
+#endif
+
+s32 func_800CBAE4(Vec3f* v0, Vec3f* v1, Vec3f* v2) {
+    s32 ret = 0;
+
+    if ((v2->x + v2->y + v2->z) < (v0->x + v0->y + v0->z)) {
+        ret = 0x01;
+    }
+
+    if ((-v1->x + v2->y + v2->z) < (-v0->x + v0->y + v0->z)) {
+        ret |= 0x02;
+    }
+
+    if ((-v1->x + v2->y - v1->z) < (-v0->x + v0->y - v0->z)) {
+        ret |= 0x04;
+    }
+
+    if ((v2->x + v2->y - v1->z) < (v0->x + v0->y - v0->z)) {
+        ret |= 0x08;
+    }
+
+    if ((v2->x - v1->y + v2->z) < (v0->x - v0->y + v0->z)) {
+        ret |= 0x10;
+    }
+
+    if ((-v1->x - v1->y + v2->z) < (-v0->x - v0->y + v0->z)) {
+        ret |= 0x20;
+    }
+
+    if ((-v1->x - v1->y + v2->z) < (-v0->x - v0->y + v0->z)) {
+        ret |= 0x40;
+    }
+
+    if ((-v1->x - v1->y - v1->z) < (-v0->x - v0->y - v0->z)) {
+        ret |= 0x80;
+    }
+    return ret;
+}
+
+s32 func_800CBC60(Vec3f* v0, Vec3f* v1, Vec3f* v2, Vec3f* v3) {
+    s32 flags[2];
```

Comment:
Is there a reason or hint to have this as an array instead of separate variables?

Reply:
This doesn't match unless it's an array.